src/org/ods/orchestration/BuildStage.groovy

```diff
@@ -103,17 +112,140 @@ class BuildStage extends Stage {
+                    def issueKey = createOrUpdateSecurityVulnerabilityIssue(
+                        vulerabilityMap.name,
+                        jiraComponentId,
+                        buildSecurityVulnerabilityIssueDescription(
+                            vulerabilityMap,
+                            repo.data.openshift.gitUrl,
+                            repo.data.openshift.gitBranch,
+                            repo.data.openshift.repoName,
+                            repo.data.openshift.nexusReportLink))
+                    securityVulnerabilityIssueKeys.add(issueKey)
+                }
+            }
+        } catch (JiraNotPresentException e) {
+            project.logger.warn(e.getMessage())
+            return []
+        }
+        return securityVulnerabilityIssueKeys
+    }
+
+    String createOrUpdateSecurityVulnerabilityIssue(String vulnerabilityName, String jiraComponentId,
+                                                    String description) {
+        if (!project.jiraUseCase || !project.jiraUseCase.jira) {
+            throw new JiraNotPresentException("JiraUseCase not present, cannot create security vulnerability issue.")
+        }
+
+        def issueSummary =  SECURITY_VULNERABILITY_ISSUE_SUMMARY.replace(VULNERABILITY_NAME_PLACEHOLDER,
+            vulnerabilityName)
+
+        def fixVersion = null
+        if (project.isVersioningEnabled) {
+            fixVersion = project.getVersionName()
+        }
+        def fullJiraComponentName = JIRA_COMPONENT_TECHNOLOGY_PREFIX + jiraComponentId
+
+        List securityVulnerabilityIssues = project?.jiraUseCase?.jira?.loadJiraSecurityVulnerabilityIssues(issueSummary,
+            fixVersion, fullJiraComponentName, project.jiraProjectKey)
+        if (securityVulnerabilityIssues?.size() >= 1) { // Transition the issue to "TO DO" state
+            transitionIssueToToDo(securityVulnerabilityIssues.get(0).id)
+            return (securityVulnerabilityIssues.get(0) as Map)?.key
+        } else { // Create the issue
+            return (createIssueTypeSecurityVulnerability(fixVersion, fullJiraComponentName,
+                SECURITY_VULNERABILITY_ISSUE_PRIORITY, projectKey: project.jiraProjectKey, summary: issueSummary,
+                description: description)
+                as Map)?.key
+        }
+    }
+
+    Map createIssueTypeSecurityVulnerability(Map args, String fixVersion = null, String component = null,
+                                             String priority = null) {
+        return project?.jiraUseCase?.jira?.createIssue(fixVersion, component, priority, summary: args.summary,
+            type: "Security Vulnerability", projectKey: args.projectKey, description: args.description)
+    }
+
+
+    void transitionIssueToToDo(String issueId) {
+        int maxAttemps = 10;
+        while (maxAttemps-- > 0) {
+            def possibleTransitions = project?.jiraUseCase?.jira?.getTransitions(issueId)
```

Comment:
To be on the safe side, we could directly check the issue status instead of inferring it from the transitions allowed. Here, we could call to api/2/issue with the parameter expand transitions. Then, we could check whether the status is to-do and, if not, execute one of the available transitions. The total number of calls would be the same.

Reply:
Done.